packages/server/api/src/app/flows/step-run/step-run-progress.handler.ts

```diff
@@ -0,0 +1,103 @@
+import { exceptionHandler } from '@activepieces/server-shared'
+import { 
+    FlowRunId,
+    isNil, 
+    PauseType, 
+    StepOutput, 
+    StepOutputStatus,
+    StepRunResponse,
+} from '@activepieces/shared'
+import { FastifyBaseLogger } from 'fastify'
+import { engineResponseWatcher } from '../../workers/engine-response-watcher'
+import { flowRunService } from '../flow-run/flow-run-service'
+
+export const stepRunProgressHandler = {
+    async notifyStepFinished(params: NotifyStepFinishedParams): Promise<void> {
+        const { runId, logger } = params
+        try {
+            const flowRun = await getFlowRun(runId, logger)
+            
+            const stepOutput = flowRun.stepNameToTest 
+                ? findStepOutputInNestedStructure(flowRun.steps, flowRun.stepNameToTest)
+                : undefined
+
+            const shouldStopProcessingStepCompletion = isNil(stepOutput) || [StepOutputStatus.RUNNING, StepOutputStatus.PAUSED].includes(stepOutput.status)
+            if (shouldStopProcessingStepCompletion) {
+                return
+            }
+
+            if (flowRun.pauseMetadata?.type !== PauseType.WEBHOOK) {
+                return
+            }
+
+            const { testCallbackRequestId, handlerId } = flowRun.pauseMetadata
+            if (isNil(handlerId) || isNil(testCallbackRequestId)) {
+                return
+            }
+
+            const response = createStepRunResponse(stepOutput, testCallbackRequestId)
+            await engineResponseWatcher(logger).publish(testCallbackRequestId, handlerId, response)
+        }
+        catch (error) {
+            exceptionHandler.handle(error, logger)
+        }
+    },
+}
+
+async function getFlowRun(runId: FlowRunId, logger: FastifyBaseLogger) {
+    let flowRun = await flowRunService(logger).getOnePopulatedOrThrow({
+        id: runId,
+        projectId: undefined,
+    })
+    
+    if (!isNil(flowRun.parentRunId)) {
+        flowRun = await flowRunService(logger).getOnePopulatedOrThrow({
+            id: flowRun.parentRunId,
+            projectId: undefined,
+        })
+    }
+    
+    return flowRun
+}
+
+function createStepRunResponse(stepOutput: StepOutput, testCallbackRequestId: string): StepRunResponse {
+    const isSuccess = stepOutput.status === StepOutputStatus.SUCCEEDED
+    
+    return {
+        id: testCallbackRequestId,
+        success: isSuccess,
+        input: stepOutput.input ?? {},
+        output: stepOutput.output ?? {},
+        standardError: isSuccess ? '' : (stepOutput.errorMessage as string),
+        standardOutput: '',
+    }
+}
+
+function findStepOutputInNestedStructure(data: unknown, stepName: string): StepOutput | undefined {
```

Comment:
This is hardcoded, I would believe it should live closer where it belongs otherwise it will break within a month 

https://github.com/activepieces/activepieces/blob/e83261a250afaea0c3d280a5f8acb25e5974fdf3/packages/shared/src/lib/flow-run/execution/step-output.ts#L4

Also I think it should understand the step type and based on that type, It continue searching.

My suggestion to make file in the shared closer to where it live, write a proper function and reuse it in the frontend and here.